include/nbl/builtin/hlsl/shapes/beziers.hlsl

```diff
@@ -133,610 +198,205 @@ namespace shapes
+            // Reducing Quartic to Cubic Solution
+            float_t kk = 1.0 / dot(A,A);
+            float_t kx = kk * dot(Bdiv2,A);
+            float_t ky = kk * (2.0*dot(Bdiv2,Bdiv2)+dot(CsubPos,A)) / 3.0;
+            float_t kz = kk * dot(CsubPos,Bdiv2);      
 
-                num = solve_quadric(quad_coeffs, s.xy);
+            float2_t res;
 
-                quad_coeffs[0]= z + u;
-                quad_coeffs[1] = q < 0. ? v : -v;
+            // Cardano's Solution to resolvent cubic of the form: y^3 + 3py + q = 0
+            // where it was initially of the form x^3 + ax^2 + bx + c = 0 and x was replaced by y - a/3
+            // so a/3 needs to be subtracted from the solution to the first form to get the actual solution
+            float_t p = ky - kx*kx;
+            float_t p3 = p*p*p;
+            float_t q = kx*(2.0*kx*kx - 3.0*ky) + kz;
+            float_t h = q*q + 4.0*p3;
 
-                float2 tmp=float2(1e38,1e38);
-                int old_num=num;
+            if(h >= 0.0) 
+            { 
+                h = sqrt(h);
+                float2_t x = (float2_t(h, -h) - q) / 2.0;
 
-                num += solve_quadric(quad_coeffs, tmp);
-                if(old_num!=num)
+                // Solving Catastrophic Cancellation when h and q are close (when p is near 0)
+                if(abs(abs(h/q) - 1.0) < 0.0001)
                 {
-                    if(old_num == 0)
-                    {
-                        s[0] = tmp[0];
-                        s[1] = tmp[1];
-                    }
-                    else
-                    {
-                        //old_num == 2
-                        s[2] = tmp[0];
-                        s[3] = tmp[1];
-                    }
+                   // Approximation of x where h and q are close with no carastrophic cancellation
+                   // Derivation (for curious minds) -> h=√(q²+4p³)=q·√(1+4p³/q²)=q·√(1+w)
+                      // Now let's do linear taylor expansion of √(1+x) to approximate √(1+w)
+                      // Taylor expansion at 0 -> f(x)=f(0)+f'(0)*(x) = 1+½x 
+                      // So √(1+w) can be approximated by 1+½w
+                      // Simplifying later and the fact that w=4p³/q will result in the following.
+                   x = float2_t(p3/q, -q - p3/q);
                 }
-            }
-
-            /* resubstitute */
 
-            float sub = 1./4. * a;
-
-            /* single halley iteration to fix cancellation */
-            for(int i=0;i<4;i+=2){
-                if(i < num){
-                    s[i] -= sub;
-                    s[i] = halley_iteration4(coeffs,s[i]);
-
-                    s[i+1] -= sub;
-                    s[i+1] = halley_iteration4(coeffs,s[i+1]);
+                float2_t uv = sign(x)*pow(abs(x), float2_t(1.0/3.0,1.0/3.0));
+                float2_t t = uv.x + uv.y - kx;
+                t = clamp( t, 0.0, 1.0 );
+                t = clipper(t.x);
```

Comment:
seems like "max distance optimization" is still on your TODO
because before clipping you should first figure out if it's even worth it, if not then set returned distance to some large value ensuring the pixel won't get drawn

Reply:
done